bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/internal/community/CommunityBundleAddonHandler.java

```diff
@@ -0,0 +1,182 @@
+
+    private final Logger logger = LoggerFactory.getLogger(CommunityBundleAddonHandler.class);
+
+    private Map<String, Long> installedBundles;
+
+    private BundleContext bundleContext;
+
+    @Activate
+    protected void activate(BundleContext bundleContext, Map<String, Object> config) {
+        this.bundleContext = bundleContext;
+        installedBundles = loadInstalledBundlesMap();
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        this.installedBundles = null;
+        this.bundleContext = null;
+    }
+
+    @Override
+    public boolean supports(String type, String contentType) {
+        // we support only certain extension types, and only as pure OSGi bundles
+        return SUPPORTED_EXT_TYPES.contains(type) && contentType.equals(BUNDLE_CONTENTTYPE);
+    }
+
+    @Override
+    public boolean isInstalled(String id) {
+        return installedBundles.containsKey(id);
+    }
+
+    @Override
+    public void install(Addon addon) throws MarketplaceHandlerException {
+        try {
+            String url = (String) addon.getProperties().get(JAR_DOWNLOAD_URL_PROPERTY);
+            Bundle bundle = bundleContext.installBundle(url);
+            try {
+                bundle.start();
+            } catch (BundleException e) {
+                logger.warn("Installed bundle, but failed to start it: {}", e.getMessage());
+            }
+            installedBundles.put(addon.getId(), bundle.getBundleId());
+            persistInstalledBundlesMap(installedBundles);
+        } catch (BundleException e) {
+            logger.debug("Failed to install bundle from marketplace.", e);
+            throw new MarketplaceHandlerException("Bundle cannot be installed: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void uninstall(Addon addon) throws MarketplaceHandlerException {
+        Long id = installedBundles.get(addon.getId());
+        if (id != null) {
+            Bundle bundle = bundleContext.getBundle(id);
+            if (bundle != null) {
+                try {
+                    bundle.stop();
+                    bundle.uninstall();
+                    installedBundles.remove(addon.getId());
+                    persistInstalledBundlesMap(installedBundles);
+                } catch (BundleException e) {
+                    throw new MarketplaceHandlerException("Failed deinstalling bundle: " + e.getMessage());
+                }
+            } else {
+                // we do not have such a bundle, so let's remove it from our internal map
+                installedBundles.remove(addon.getId());
+                persistInstalledBundlesMap(installedBundles);
+                throw new MarketplaceHandlerException("Id not known.");
+            }
+        } else {
+            throw new MarketplaceHandlerException("Id not known.");
+        }
+    }
+
+    private Map<String, Long> loadInstalledBundlesMap() {
+        File dataFile = bundleContext.getDataFile(BUNDLE_FILE);
```

Comment:
I do not remember if this came from my initial implementation, but afaik, the datafile of bundles will be wiped upon an openHAB update or the infamous "delete cache" that many people are doing regularly.
I guess we should therefore rather store this file in the USER_DATA folder along with our other stuff. Wdyt?

Reply:
It was, and you mentioned that particular problem in the Javadoc even: https://github.com/openhab/openhab2-addons/blob/2.5.x/bundles/org.openhab.extensionservice.marketplace/src/main/java/org/openhab/extensionservice/marketplace/internal/BundleExtensionHandler.java#L39-L49

So I moved the code over mostly untouched. I don't know if you had a good reason at the time not to put that file in USER_DATA but if you deem it's okay now, we can do that.

Reply:
Looks indeed like a temporary solution that was done as a quick hack that then lasts forever. :-)
Yeah, I guess it makes sense to replace this by storing the file in the USER_DATA folder.

Reply:
In fact you might have done that on purpose:
> The information, which installed bundle corresponds to which extension is written to a file in the bundle's data store. It is therefore wiped ***together with the bundles*** upon an OSGi "clean".

If a cleanup or upgrade wipes the bundles anyway, it makes sense for that file to be cleaned as well, if there's no mechanism to re-install the bundles from that file who have gone.

Reply:
Hm, ok, so with an openHAB update all installed add-ons from the marketplace get removed as well and therefore the information on "what is installed" is outdated and can be wiped as well. But aren't we then lacking an information of what the user wants to have installed, so that it gets reinstalled automatically? Similar to the addons.cfg for the official add-ons? Or Are is the information for marketplace addons stored in there as well? In that case, all would be fine indeed.

Reply:
Quoting your own words 😄  (https://community.openhab.org/t/distributing-bindings-through-the-iot-marketplace/24491):

> Note 2: The jars are installed in the OSGi cache - which means that they are lost upon an openHAB distro upgrade (which cleans the cache) - you will have to manually re-install then from the marketplace. As the main use case for the start will be for testing pre-releases, I think this is acceptable. For the future, I also plan an improved mechanism, which would download the jars to the addons folder (or something similar), so that they are kept over an upgrade as well.

So it's definitely an inherent problem with this implementation, what you're suggesting could be a solution (downloading them to `addons` on install instead, keeping track of the files there to maintain the installed status and removing them on uninstall). Otherwise I don't really yet know how a reinstallation of bundles that are marked as installed in the file but missing could work.
The approach with the CSV file is problematic since entries in that file are like
```
marketplace:123568;316
```
316 being the bundle ID, which I guess is not even reliable after a cleanup.

Reply:
I didn't look in the code, why do you need the bundle id? You could check if the BSN is in the list of installed bundles.

Reply:
Digging more into the API it appears using https://docs.osgi.org/javadoc/osgi.core/8.0.0/org/osgi/framework/BundleContext.html#installBundle-java.lang.String-java.io.InputStream-
we can set any "location identifier" to any arbitrary string, not necessarily the actual source URL, which can then be set to the add-on ID like `marketplace:12345`.
Then `isInstalled(String id)` becomes simply `bundleContext.getBundle(id) != null`
(https://docs.osgi.org/javadoc/osgi.core/8.0.0/org/osgi/framework/BundleContext.html#getBundle-java.lang.String-)

And therefore no need to keep a map in a CSV file at all, see https://github.com/openhab/openhab-core/pull/2405/commits/69c97de5467ce64aa9eab23144e3fc8fd740fb6b. It appears to work well.
Note that this doesn't solve the problem of the user having to re-install marketplace bundles after an update or cache cleaning.

Reply:
You could store a list (e.g. comma-separated) with the `ConfigurationAdmin` or use the `StorageService`. The latter stores the information in the user-data directory, which is not cleared.

Reply:
The question is whether we want to only store the information about which add-ons are installed (i.e. the file with the ids) or if we want to keep the add-on jars themselves and automatically re-install them.
The former (ids only) is more aligned with how we handle the official add-ons. But for the marketplace, we do not offer an offline version as we do with the add-ons kar, so people would have to rely on the marketplace and that their add-ons keep being available there, which somehow feels a bit risky to me.
What we could maybe do is to download them to a dedicated folder in USER_DATA and install them from there. This way, we can reinstall them easily after an openHAB update. Wdyt?

Reply:
Initially I thought about re-installing them from their online location (or ignore it if the marketplace entry doesn't exist anymore) but keeping them locally makes sense.

> What we could maybe do is to download them to a dedicated folder in USER_DATA and install them from there. This way, we can reinstall them easily after an openHAB update. Wdyt?

In other terms - another cache, separate from OSGi's but one that would be preserved; it would be "elegant" to put it in the cache folder alongside OSGi's (`$USERDATA/cache/marketplace` or similar) but we can't because the upgrade scripts and clean cache "procedure" effectively delete the whole of `$USERDATA/cache`; not only `$USERDATA/cache/org.eclipse.osgi` unfortunately.

So maybe a structure like: `$USERDATA/marketplace/bundles/12345/<originalname>.jar` (12345 being the id of the marketplace add-on). That way when the service activates, it can iterate over `$USERDATA/marketplace/bundles/*` and check for its actual installation status (by locating a bundle whose location id is `marketplace:<id>` as above). If it turns out there is no matching bundle then reinstall it from the .jar stored locally.
When the add-on is removed, the folder and .jar should be removed, of course, so that no re-installation is attempted. I believe having only this folder structure and no file to edit makes it easy enough to understand for admins.

Reply:
You should take into account that "older" versions might not be compatible on upgrade since there is no fixed versioning scheme for 3rd party addons.

It might be more difficult to validate compatibility with a local cache than with a remote repository.

Reply:
>You should take into account that "older" versions might not be compatible on upgrade since there is no fixed versioning scheme for 3rd party addons.
> It might be more difficult to validate compatibility with a local cache than with a remote repository.

That's a valid point for updates; on the other hand, sometimes users would want to stick to the version they know works and not having it updated without their knowledge from the online source because they busted their OSGi cache.
I would suggest removing from the local cache any bundle/add-on if it doesn't start during reinstallation (and also uninstalling it). That way it would simply be completely uninstalled again and users can go to the marketplace UI and reinstall it from the source.

Reply:
Implemented in https://github.com/openhab/openhab-core/pull/2405/commits/c00b2c6c4d5565c09e398b6809d0340d50508bfd.

```
____________________________
Welcome to Apache Felix Gogo

g! 11:11:19.007 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Time zone set to 'Europe/Paris'.
11:11:19.361 [INFO ] [rketplace.MarketplaceBundleInstaller] - Reinstalling missing marketplace bundle: marketplace:123568
11:11:24.130 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'demo.items'
11:11:24.742 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'mapdb.persist'
11:11:25.027 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'demo.sitemap'
11:11:25.249 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'demo.things'
11:11:27.530 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'demo.rules'
11:11:30.425 [INFO ] [org.openhab.ui.internal.UIService   ] - Started UI on port 8080
```

Reply:
Uninstalling immediately if it does not start might not be the best way. It could be that a bundle depends on a not yet installed core-feature (e.g. openhab-transport-upnp is quite common as a dependency) but will resolve later (e.g. if you install a distro addon feature that also depends on the same core feature). Bundle installation order is not fixed, so you cannot rely on it. Maybe skip the deinstallation issue at the moment, bundles that don't start because of dependency issue do not hurt.

And I would still suggest to (at least additionally) store the bsn (or the GAV coordinates). A later version of the marketplace (not needed at the moment) might check the online repository to see if an updated bundle is available if the one present in cache is not compatible.

Reply:
> Maybe skip the deinstallation issue at the moment, bundles that don't start because of dependency issue do not hurt.

Ok, fair point. So then it's back to the admin to diagnose why an add-on hasn't resolved/doesn't start after an upgrade. The add-on will still be marked as installed in the marketplace UI (it is, after all, installed, but just perhaps needs an upgrade too). In any case, uninstalling and reinstalling from there should update it to the latest version.

Reply:
Regarding future openHAB versions it might be a good idea to add something like 

```
    "compatibility": [
        "3.1.*",
        "3.2.*"
    ]
```

to the `properties` section. That way you can filter incompatible versions (you can get the current version at runtime with something like `String ohcVersion = FrameworkUtil.getBundle(org.openhab.core.OpenHAB.class).getVersion().toString();`. With that an author might provide different versions for different versions of openHAB core/distribution.

Of course this is not necessary in the first place, since there will only be 3.2.x, but at least the format of the JSON should allow it.

Reply:
I agree, some version management makes sense mid-term. Not mandatory for this PR, but if marketplace authors would already provide such properties, we would be safe in the future.

Reply:
I agree that version/dependency management is important, but IMHO we shouldn't go too far and reinvent the wheel in this case - those are better managed with e.g. Karaf features, this is what they are for.

So while I don't know what the use cases for the marketplace will be in the end, I feel that as a first step it should only remain a stepping stone for bindings that are not just ready for the distribution yet - instead of developers telling users "here a .jar that I built, it might or might not work for you, but here you go, drop it in your addons folder", the "drop it in your addons folder" part is replaced by a click of a button.
Of course improvements can be made down the line, like instead of linking to a .jar file in their forum post, authors would have the option of linking to a manifest XML or JSON file that would describe the versions of their add-on, and the versions of openHAB they are compatible with; but then again, we are reinventing things that are better managed when the add-on is distributed as a karaf feature, and this should be what contributors should ultimately, and ideally, strive to.

Reply:
If we have version information only within the bundle or in Karaf features, you would only be able to tell at installation time whether it fits to your openHAB version or not. Some developers provide already now 2 jars, one for the latest stable (3.1) and one for snapshot (3.2) - so imho we will need a way for them to directly mark it accordingly in the community, so that only one of these two versions is offered within openHAB (depending on its version). And we will quickly also see older marketplace entries that might not be compatible with openHAB 3.3, 3.4, etc. anymore, while working nicely with e.g. 3.1.
TL;DR: I don't think we'll get around it...